options/posix/generic/posix-file-io.cpp

```diff
@@ -30,58 +21,160 @@ int mem_file::determine_bufmode(buffer_mode *mode) {
 	return 0;
 }
 
-int mem_file::io_read(char *, size_t, size_t *) {
-	return EINVAL;
+int mem_file::io_read(char *buffer, size_t max_size, size_t *actual_size) {
+	if ((_pos >= 0 && _pos >= _max_size) || !max_size) {
+		*actual_size = 0;
+		return 0;
+	}
+
+	size_t bytes_read = std::min(size_t(_max_size - _pos), max_size);
+	memcpy(buffer, _buffer().data() + _pos, bytes_read);
+	_pos += bytes_read;
+	*actual_size = bytes_read;
+	return 0;
 }
 
-int mem_file::io_write(const char *buffer, size_t max_size, size_t *actual_size) {
-	if (_pos + max_size >= _buf.size()) {
+memstream_mem_file::memstream_mem_file(char **ptr, size_t *sizeloc, int flags, void (*do_dispose)(abstract_file *))
+: mem_file{flags, do_dispose}, _bufloc{ptr}, _sizeloc{sizeloc} { }
+
+
+int memstream_mem_file::close() {
+	_update_ptrs();
+	_buf.detach();
+
+	return 0;
+}
+
+int memstream_mem_file::io_write(const char *buffer, size_t max_size, size_t *actual_size) {
+	if (_pos + max_size >= _buffer_size()) {
 		_buf.resize(_pos + max_size + 1, '\0');
 		_update_ptrs();
 	}
 
-	memcpy(_buf.data() + _pos, buffer, max_size);
+	size_t bytes_write = std::min(_buffer_size() - _pos, max_size);
+	memcpy(_buffer().data() + _pos, buffer, bytes_write);
 	_pos += max_size;
 	*actual_size = max_size;
+
 	return 0;
 }
 
-int mem_file::io_seek(off_t offset, int whence, off_t *new_offset) {
+int memstream_mem_file::io_seek(off_t offset, int whence, off_t *new_offset) {
 	switch (whence) {
 		case SEEK_SET:
 			_pos = offset;
-			if (_pos >= _buf.size()) {
+			if (_pos >= 0 && size_t(_pos) >= _buffer_size()) {
 				_buf.resize(_pos + 1, '\0');
 				_update_ptrs();
 			}
 			*new_offset = _pos;
 			break;
 		case SEEK_CUR:
 			_pos += offset;
-			if (_pos >= _buf.size()) {
+			if (_pos >= 0 && size_t(_pos) >= _buffer_size()) {
 				_buf.resize(_pos + 1, '\0');
 				_update_ptrs();
 			}
 			*new_offset = _pos;
 			break;
 		case SEEK_END:
-			_pos = _buf.size() ? _buf.size() - 1 + offset : _buf.size() + offset;
+			_pos = _buffer_size() ? _buffer_size() - 1 + offset : _buffer_size() + offset;
 			_buf.resize(_pos + 1, '\0');
 			_update_ptrs();
 			*new_offset = _pos;
 			break;
 		default:
-			__ensure(!"Unknown whence value passed!");
-			__builtin_unreachable();
+			return EINVAL;
 	}
 	return 0;
 }
 
-void mem_file::_update_ptrs() {
+void memstream_mem_file::_update_ptrs() {
 	*_bufloc = _buf.data();
 	*_sizeloc = _buf.size() - 1;
 }
 
+fmemopen_mem_file::fmemopen_mem_file(void *in_buf, size_t size, int flags, void (*do_dispose)(abstract_file *))
+: mem_file{flags, do_dispose}, _inBuffer{in_buf}, _inBufferSize{size} {
+	if(!_inBuffer) {
+		_inBuffer = getAllocator().allocate(size);
```

Comment:
Do we properly handle the case when ``size`` is 0? On glibc this will cause the first read to return EOF, but im not sure that happens here (don't we just crash?) This behavior should obviously be tested ofc.

Reply:
We correctly implement it, although we could also return EINVAL and still be compliant with POSIX. I added a test case.

Reply:
Fun side note: GHA's glibc actually returns EINVAL with size 0.